### PR TITLE
Lecture 2

### PR DESCRIPTION
**Pull request by HidJaba-hub**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 1]
```diff
@@ -46,20 +51,41 @@ public Optional<Product> getProductById(Long id) {
     }
 
     @Override
-    public List<Product> findProducts() {
+    public List<Product> findProducts(SortField sortField, SortOrder sortOrder, String query) {
         readLock.lock();
         try {
+            Comparator<Product> comparator = setComparator(sortField, sortOrder, query);
             return products.stream()
+                    .filter(product -> StringUtils.isEmpty(query) || StringChecker.calculateStringSimilarity(product.getDescription(), query) > 0)
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: too long string

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 2]
```diff
@@ -46,20 +51,41 @@ public Optional<Product> getProductById(Long id) {
...
+                    .sorted(comparator)
                     .collect(Collectors.toList());
         } finally {
             readLock.unlock();
         }
 
     }
 
+    private Comparator<Product> setQueryComparator(String query) {
+        return Comparator.comparing(Product::getDescription, (s1, s2) -> Double.compare(StringChecker.calculateStringSimilarity(s2, query), StringChecker.calculateStringSimilarity(s1, query)));
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: too long string

File: src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java [hunk 3]
```diff
@@ -1,11 +1,18 @@
 package com.es.phoneshop.exception;
 
 public class ProductDefinitionException extends RuntimeException {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java]: maybe let's rename to smth like ProductNotFoundException? not sure about "Definition" part in naming

File: src/main/java/com/es/phoneshop/model/entity/Product.java [hunk 4]
```diff
@@ -28,17 +28,20 @@ public class Product {
...
-    public Product(){
-        this.id=UUID.randomUUID().getMostSignificantBits();
+    private List<PriceHistory> priceHistory;
+
+    public Product() {
+        this.id = UUID.randomUUID().getMostSignificantBits();
     }
 
-    public Product(String code, String description, BigDecimal price, Currency currency, int stock, String imageUrl) {
+    public Product(String code, String description, BigDecimal price, Currency currency, int stock, String imageUrl, List<PriceHistory> priceHistoryList) {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/model/entity/Product.java]: too long

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Currency;
+import java.util.Date;
+import java.util.List;
+
+public class DemoDataServletContextListener implements ServletContextListener {
+    //as i understand i need to test this class too, but i have no idea at all how to do it
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: you can try to test contextInitialized()

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        productService.saveProduct(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg", generateRandomPriceHistory()));
+        productService.saveProduct(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg", generateRandomPriceHistory()));
+        productService.saveProduct(new Product("htces4g", "HTC EVO Shift 4G", new BigDecimal(320), usd, 3, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/HTC/HTC%20EVO%20Shift%204G.jpg", generateRandomPriceHistory()));
+        productService.saveProduct(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg", generateRandomPriceHistory()));
+    }
+
+    List<PriceHistory> generateRandomPriceHistory() {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: for me price history doesn't work

Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: private

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+        for (int i = 0; i < priceHistoryLength; i++) {
+            Date date = Date.from(createRandomDate().atStartOfDay(ZoneId.systemDefault()).toInstant());
+            priceHistoryList.add(new PriceHistory(date, usd, new BigDecimal(createRandomIntBetween(100, 5000))));
+        }
+        priceHistoryList.sort(Comparator.comparing(PriceHistory::getDate));
+        return priceHistoryList;
+    }
+
+    public static int createRandomIntBetween(int start, int end) {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: private

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        return start + (int) Math.round(Math.random() * (end - start));
+    }
+
+    public static LocalDate createRandomDate() {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: private

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java (hunk 7 above)
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: why static?

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java (hunk 8 above)
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: why static?

File: src/main/webapp/WEB-INF/web.xml [hunk 9]
```diff
@@ -1,18 +1,48 @@
...
+
+    <error-page>
+        <location>/WEB-INF/pages/errorProductDefinitionException.jsp</location>
+        <exception-type>com.es.phoneshop.exception.ProductDefinitionException</exception-type>
+    </error-page>
+
+    <error-page>
+        <location>/WEB-INF/pages/error.jsp</location>
+        <error-code>500</error-code>
+
```
Comment (esSoftPalinaTsi) [inline src/main/webapp/WEB-INF/web.xml:61]: remove line

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    @Mock
+    private RequestDispatcher requestDispatcher;
+    @Mock
+    private ServletConfig config;
+
+    @Before
+    public void setup() throws ServletException {
+        servlet.init(config);
+    }
+
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java:43]: test for positive case

File: src/main/java/com/es/phoneshop/StringChecker.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+            return 0;
+        }
+
+        Set<String> criteriaWords = Arrays.stream(criteriaString.toLowerCase().split("\\s+"))
+                .collect(Collectors.toSet());
+
+        double wordsCount = Arrays.stream(sourceString.toLowerCase().split("\\s+"))
+                .filter(criteriaWords::contains)
+                .count();
+        return wordsCount / sourceString.length();
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/StringChecker.java:22]: why do we need such division? didn't get

File: src/main/java/com/es/phoneshop/service/CustomProductService.java [hunk 12]
```diff
@@ -39,13 +43,13 @@ public Product getProductById(Long id) {
         if (optionalProduct.isPresent()) {
             return optionalProduct.get();
         } else {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/CustomProductService.java]: orElseThrow()

File: src/main/webapp/WEB-INF/web.xml (hunk 9 above)
Comment (esSoftPalinaTsi) [inline src/main/webapp/WEB-INF/web.xml:61]: and + for 404

File: src/main/java/com/es/phoneshop/StringChecker.java (hunk 11 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/StringChecker.java:22]: this is necessary for relevance. because if we have for example "samsung s II" and "samusng s lII", and our query is "samsung S", it will be more similar to the 1st than to the 2nd because this coefficient count how similar our request is to our string(in this case we have both 2 matched words, but we must decide what string will be closer by simply dividing it by length). So based on this number, if the query is not empty, we sort our list according to relevance

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java (hunk 8 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: i supposef thst i could use this method in another classes, but i think in this situation i mustn't be static

File: src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java (hunk 3 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java]: That because i use this exception for 2 cases: 1. for product not found, 2. for product is null. Maybe i need to create 2 custom exceptions both for 2 cases?

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java (hunk 6 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: this generation is definitely working correctly. the problem i suppose in displaying this part

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DemoDataServletContextListenerTest {
+    ProductDao productDao = CustomProductDao.getInstance();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: all variables need to be private

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    DemoDataServletContextListener demoDataServletContextListener = new DemoDataServletContextListener();
+    @Mock
+    ServletContext context;
+    @Mock
+    ServletContextEvent event;
+
+    @Before
+    public void setup() {
+        when(event.getServletContext()).thenReturn(context);
+        when(context.getInitParameter("insertDemoData")).thenReturn("true");
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: and + one negative case when insertDemoData = false

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Test
+    public void initParameterTest() {//is it properly named?
+        boolean insertDemoData = Boolean.valueOf(event.getServletContext().getInitParameter("insertDemoData"));
+
+        assertTrue(insertDemoData);
+    }
+
+    @Test
+    public void testContextInitialized() {
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: givenInsertDemoDataTrue_whenContextInitialized_thenInsertData

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    ServletContextEvent event;
+
+    @Before
+    public void setup() {
+        when(event.getServletContext()).thenReturn(context);
+        when(context.getInitParameter("insertDemoData")).thenReturn("true");
+    }
+
+    @Test
+    public void initParameterTest() {//is it properly named?
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: no need for this test at all, your listener contains not empty implementation only for contextInitialized()

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        boolean insertDemoData = Boolean.valueOf(event.getServletContext().getInitParameter("insertDemoData"));
+
+        assertTrue(insertDemoData);
+    }
+
+    @Test
+    public void testContextInitialized() {
+        demoDataServletContextListener.contextInitialized(event);
+
+        assertNotNull(productDao.getProducts());
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: no need to call multiple times productDao.getProducts(), please save it to variable and remember this everywhere (for all services, tests, daos, etc)

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        when(request.getPathInfo()).thenReturn("/" + productId);
+
+        servlet.doGet(request, response);
+
+        verify(requestDispatcher).forward(request, response);
+        verify(request).setAttribute(eq("product"), any());
+    }
+
+    @Test
+    public void givenProduct_whenGetRequest_thenVerify() throws ServletException, IOException {
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: thenVerify - verify what? When I'm reading the test name I should have clear understanding of the test case, what it's checking and what is the result of the check

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 19]
```diff
@@ -46,20 +53,59 @@ public Optional<Product> getProductById(Long id) {
...
+        return Comparator.comparing(Product::getDescription,
+                (s1, s2) -> Double.compare(
+                        StringChecker.calculateStringSimilarity(s2, query),
+                        StringChecker.calculateStringSimilarity(s1, query)
+                )
+        );
+    }
+
+    private Comparator<Product> setFieldComparator(SortField sortField, String query, Comparator<Product> comparator) {
+        if (sortField == null) return comparator;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: formatting

File: src/main/java/com/es/phoneshop/DAO/ProductDao.java [hunk 20]
```diff
@@ -8,13 +10,15 @@
...
     void save(Product product);
 
     void delete(Long id);
 
     void changeChosenState(Product product, boolean state);
-    void setProducts(List<Product> products);
+
     List<Product> getProducts();
+
+    void setProducts(List<Product> products);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/ProductDao.java]: setter should not be present in dao interface
interface knows nothing about the DAO implementation and that DAO uses list as a storage

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 21]
```diff
@@ -23,8 +26,15 @@ public void init(ServletConfig servletConfig) throws ServletException {
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products", productService.getProducts());
-        productService.getProducts().forEach(product -> productService.changeState(product, false));
+        String sort = request.getParameter("sort");
+        String order = request.getParameter("order");
+        String query = request.getParameter("query");
+
+        SortField sortField = StringUtils.isEmpty(sort) ? null : SortField.valueOf(sort.toUpperCase());
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: why do we need null here? Just not call sorting with this param/set it as empty string
In general it's bad idea to hardcore something with null-s, because some time later it can lead to unexpected NPE-s

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Currency;
+import java.util.Date;
+import java.util.List;
+
+public class DemoDataServletContextListener implements ServletContextListener {
+    private final Currency usd = Currency.getInstance("USD");
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java:21]: line after class definition

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 23]
```diff
@@ -18,16 +22,19 @@ public class CustomProductDao implements ProductDao {
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
     private final Lock readLock = lock.readLock();
     private final Lock writeLock = lock.writeLock();
+    private final Comparator<Product> defaultComparator = Comparator.comparing(Product::getDescription);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: by default there is no sort
this will increase the pressure on server
when 1mln of users tries to go on PLP at  one time, they are interested only in products
and only when some of them need sorting, sorting should happen
we should be careful with resources

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 24]
```diff
@@ -46,20 +53,59 @@ public Optional<Product> getProductById(Long id) {
...
+                            setQueryComparator(query, defaultComparator)));
+            // i created this nested system of comparators that are independent of each other so it can be easily expanded
+            // it is also safe because as a root we have default comparator, and if all parameters will be null it will just return default comparator
             return products.stream()
+                    .filter(product -> StringUtils.isEmpty(query)
+                            || StringChecker.calculateStringSimilarity(product.getDescription(), query) > 0)
                     .filter(product -> product.getPrice() != null)
                     .filter(product -> product.getPrice().compareTo(BigDecimal.valueOf(0)) > 0)
                     .filter(product -> product.getStock() > 0)
+                    .sorted(comparator)
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: as I said - we need to sort not just for all the times, but only when user requested the sort

File: src/main/java/com/es/phoneshop/DAO/ProductDao.java [hunk 25]
```diff
@@ -8,13 +10,15 @@
...
     Optional<Product> getProductById(Long id);
 
-    List<Product> findProducts();
+    List<Product> findProducts(SortField sortField, SortOrder sortOrder, String query);
 
     void save(Product product);
 
     void delete(Long id);
 
     void changeChosenState(Product product, boolean state);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/ProductDao.java]: please remove
and clean the unnecessary attributes from POJO

File: src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java (hunk 3 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java:3]: 1223

Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java:3]: 1223123

Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java:3]: 1223

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 26]
```diff
@@ -46,20 +53,59 @@ public Optional<Product> getProductById(Long id) {
     }
 
     @Override
-    public List<Product> findProducts() {
+    public List<Product> findProducts(SortField sortField, SortOrder sortOrder, String query) {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: let's split the current method implementation for 2 methods: when the sorting/search are needed and when they are not needed
and depending on the fact if sort/search are needed implementation will look better

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+        productService = CustomProductService.getInstance();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        long productId = Long.valueOf(request.getPathInfo().substring(1));
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: what can return request.getPathInfo()?

File: src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java (hunk 3 above)
Comment (esSoftPalinaTsi) [reply src/main/java/com/es/phoneshop/exception/ProductDefinitionException.java]: got it
then let's split this 2 cases with 2 exceptions: in case of nullable argument we will throw IllegalArgumentException, in case of not found product - your custom one

File: src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java (hunk 6 above)
Comment (esSoftPalinaTsi) [reply src/main/java/com/es/phoneshop/web/listener/DemoDataServletContextListener.java]: then please investigate and fix the issues

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java (hunk 27 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: if there is no extra path information passed, this function will return null. Or it can be extra situation where it hasn't any numbers that we trying to get. I will handle this situation

File: src/main/java/com/es/phoneshop/DAO/ProductDao.java [hunk 28]
```diff
@@ -10,15 +10,16 @@
 public interface ProductDao {
     Optional<Product> getProductById(Long id);
 
-    List<Product> findProducts(SortField sortField, SortOrder sortOrder, String query);
+    List<Product> findProducts();
+
+    List<Product> sortProducts(SortField sortField, SortOrder sortOrder, String query);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/ProductDao.java]: sortProducts() contains some query in params
method naming or it's purpose are not correct

File: src/main/java/com/es/phoneshop/model/entity/Product.java [hunk 29]
```diff
@@ -44,4 +44,9 @@ public Product(String code, String description, BigDecimal price, Currency curre
         this.imageUrl = imageUrl;
         this.priceHistory = priceHistoryList;
     }
+
+    @Override
+    public int compareTo(Product o) {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/model/entity/Product.java]: do we really need Comparable<Product> and such compareTo?

File: src/main/java/com/es/phoneshop/web/ProductSortServlet.java [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.apache.maven.shared.utils.StringUtils;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ProductSortServlet extends HttpServlet {
+    public ProductService productService;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductSortServlet.java]: one line after class definition

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 31]
```diff
@@ -7,6 +7,7 @@
 import jakarta.servlet.http.HttpServlet;
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.servlet.http.HttpServletResponse;
+import org.apache.maven.shared.utils.StringUtils;
 
 import java.io.IOException;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:13]: one line after class definition

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 32]
```diff
@@ -12,11 +14,12 @@
 import org.apache.maven.shared.utils.StringUtils;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public class ProductListPageServlet extends HttpServlet {
     public ProductService productService;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: one line after class definition

File: src/main/java/com/es/phoneshop/web/ProductSortServlet.java [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String sort = request.getParameter("sort");
+        String order = request.getParameter("order");
+        String query = request.getParameter("query");
+        List<Product> products = productService.getProductsWithSortingAndQuery(SortField.valueOf(sort.toUpperCase()), SortOrder.valueOf(order.toUpperCase()), query);
+        String json = objectMapper.writeValueAsString(products.stream()
+                                                                .map(Product::getId)
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductSortServlet.java]: formatting

File: src/test/java/com/es/phoneshop/DAO/CustomProductDaoTest.java [hunk 34]
```diff
@@ -44,14 +43,14 @@ public void initializeMocks() {
                     when(product.getStock()).thenReturn(10);
                     when(product.getPrice()).thenReturn(new BigDecimal("10.0"));
                 });
-        productDao.setProducts(mockedProducts);
+        productDao.getProducts().addAll(0, mockedProducts);
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/DAO/CustomProductDaoTest.java]: productDao.setProducts(mockedProducts)

File: src/main/java/com/es/phoneshop/model/entity/Product.java (hunk 29 above)
Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/model/entity/Product.java]: in order not to return null in comparators, but Comparator.naturalOrdering

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 35]
```diff
@@ -20,6 +20,7 @@
 
 @RunWith(MockitoJUnitRunner.class)
 public class DemoDataServletContextListenerTest {
+
     private ProductDao productDao = CustomProductDao.getInstance();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: is DAO called in listener?

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 36]
```diff
@@ -0,0 +1,79 @@
...
+    private ServletConfig config;
+    @InjectMocks
+    private CustomProductService productService;
+    @Mock
+    private ProductDao productDao;
+
+    @Before
+    public void setup() throws ServletException {
+        servlet.init(config);
+        productService = CustomProductService.getInstance();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: your productService is mocked already, why do you need getInstance()?

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+    private final ProductDetailsPageServlet servlet = new ProductDetailsPageServlet();
+    @Mock
+    private HttpServletRequest request;
+    @Mock
+    private HttpServletResponse response;
+    @Mock
+    private RequestDispatcher requestDispatcher;
+    @Mock
+    private ServletConfig config;
+    @InjectMocks
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: why we need here @InjectMocks, @Mock should be enough

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 38]
```diff
@@ -0,0 +1,79 @@
...
+        verify(requestDispatcher).forward(request, response);
+        verify(request).setAttribute(eq("product"), any());
+    }
+
+    @Test
+    public void givenProduct_whenGetRequest_thenSetProductToAttribute() throws ServletException, IOException {
+        Product product = new Product();
+        long productId = product.getId();
+        when(request.getPathInfo()).thenReturn("/" + productId);
+
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: remove line

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 39]
```diff
@@ -0,0 +1,57 @@
...
+    private ServletContext context;
+    @Mock
+    private ProductDao productDaoMock;
+    @Mock
+    private ServletContextEvent event;
+
+    @Before
+    public void setup() {
+        when(event.getServletContext()).thenReturn(context);
+
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: remove line

File: src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java [hunk 40]
```diff
@@ -0,0 +1,57 @@
...
+        verifyZeroInteractions(productDaoMock);
+    }
+
+    @Test
+    public void givenInsertDemoDataTrue_whenContextInitialized_thenInsertData() {
+        when(context.getInitParameter("insertDemoData")).thenReturn("true");
+        demoDataServletContextListener.contextInitialized(event);
+
+        List<Product> products = productDao.getProducts();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/listener/DemoDataServletContextListenerTest.java]: your listener knows nothing about DAO, it has only service
you should rely only on the resources which are defined in class which you are testing

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 41]
```diff
@@ -0,0 +1,79 @@
...
+    @Mock
+    private HttpServletResponse response;
+    @Mock
+    private RequestDispatcher requestDispatcher;
+    @Mock
+    private ServletConfig config;
+    @InjectMocks
+    private CustomProductService productService;
+    @Mock
+    private ProductDao productDao;
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: why do we need dao in servlet?

File: src/main/java/com/es/phoneshop/service/ProductService.java [hunk 42]
```diff
@@ -1,17 +1,22 @@
...
+
     Product getProductById(Long id);
 
     List<Product> getProducts();
 
-    void deleteProduct(Long id);
+    List<Product> findProductsByQuery(String query);
 
-    void changeState(Product product, boolean state);
+    List<Product> getProductsWithSortingAndQuery(SortField sortField, SortOrder sortOrder, String query);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/ProductService.java]: method namings should be consistent:  getProducts...() and findProducts...() should have single prefix

File: src/main/java/com/es/phoneshop/model/entity/Product.java (hunk 29 above)
Comment (esSoftPalinaTsi) [reply src/main/java/com/es/phoneshop/model/entity/Product.java]: we shouldn't hack POJO to make DAO working, can this be done in a different way?

File: src/main/java/com/es/phoneshop/DAO/CustomProductDao.java [hunk 43]
```diff
@@ -57,21 +63,84 @@ public List<Product> findProducts() {
...
+            return getProductList(query, comparator);
+        } finally {
+            readLock.unlock();
+        }
 
     }
 
+    private Comparator<Product> setQueryComparator(String query) {
+        if (StringUtils.isEmpty(query)) {
+            return Comparator.naturalOrder();
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: if query is empty we can simply return not sorted products from findProducts()

Comment (HidJaba-hub) [reply src/main/java/com/es/phoneshop/DAO/CustomProductDao.java]: but here the comparator returns, not the list of sorted/unsorted products

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java (hunk 41 above)
Comment (HidJaba-hub) [reply src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: to mock find product by id, because from service it won't work

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java (hunk 37 above)
Comment (HidJaba-hub) [reply src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: without inject mock product dao won't be injected into product service

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java (hunk 36 above)
Comment (HidJaba-hub) [reply src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: because of inject mock, when be inject smth we must initialize this object
